File: petblocks-bukkit-plugin/src/main/java/com/github/shynixn/petblocks/bukkit/PetBlocksPlugin.java
package com.github.shynixn.petblocks.bukkit;

import com.github.shynixn.petblocks.api.PetBlocksApi;
import com.github.shynixn.petblocks.api.business.controller.PetBlockController;
import com.github.shynixn.petblocks.api.business.service.GUIService;
import com.github.shynixn.petblocks.api.persistence.controller.PetMetaController;
import com.github.shynixn.petblocks.bukkit.logic.business.GoogleGuiceBinder;
import com.github.shynixn.petblocks.bukkit.logic.business.PetBlockManager;
import com.github.shynixn.petblocks.bukkit.logic.business.helper.UpdateUtils;
import com.github.shynixn.petblocks.bukkit.logic.business.listener.InventoryListener;
import com.github.shynixn.petblocks.bukkit.nms.NMSRegistry;
import com.github.shynixn.petblocks.bukkit.nms.VersionSupport;
import com.github.shynixn.petblocks.core.logic.business.helper.ReflectionUtils;
import com.github.shynixn.petblocks.core.logic.persistence.configuration.Config;
import com.google.inject.Guice;
import com.google.inject.Inject;
import org.apache.commons.io.IOUtils;
import org.bstats.bukkit.Metrics;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.*;
import java.lang.reflect.InvocationTargetException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Main class of the plugin PetBlocks.
 * <p>
 * Version 1.2
 * <p>
 * MIT License
 * <p>
 * Copyright (c) 2018 by Shynixn
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
public final class PetBlocksPlugin extends JavaPlugin {
    public static final String PREFIX_CONSOLE = ChatColor.AQUA + "[PetBlocks] ";
    private static final long SPIGOT_RESOURCEID = 12056;
    private static final String PLUGIN_NAME = "PetBlocks";
    private static Logger logger;
    private boolean disabled;

    private PetBlockManager petBlockManager;

    /**
     * Listeners.
     */
    @Inject
    private InventoryListener inventoryListener;

    @Inject
    private GUIService guiService;

    /**
     * Enables the plugin PetBlocks.
     */
    @Override
    public void onEnable() {
        this.saveDefaultConfig();
        logger = this.getLogger();
        if (!VersionSupport.isServerVersionSupported(PLUGIN_NAME, PREFIX_CONSOLE)) {
            this.disabled = true;
            Bukkit.getPluginManager().disablePlugin(this);
        } else {
            Bukkit.getServer().getConsoleSender().sendMessage(PREFIX_CONSOLE + ChatColor.GREEN + "Loading PetBlocks ...");
            Guice.createInjector(new GoogleGuiceBinder(this));
            Config.getInstance().reload();
            if (Config.getInstance().isMetricsEnabled()) {
                final Metrics metrics = new Metrics(this);
                metrics.addCustomChart(new Metrics.SimplePie("storage", () -> {
                    if (PetBlocksPlugin.this.getConfig().getBoolean("sql.enabled")) {
                        return "MySQL";
                    }
                    return "SQLite";
                }));
            }
            this.getServer().getScheduler().runTaskAsynchronously(this, () -> {
                try {
                    UpdateUtils.checkPluginUpToDateAndPrintMessage(SPIGOT_RESOURCEID, PREFIX_CONSOLE, PLUGIN_NAME, PetBlocksPlugin.this);
                } catch (final IOException e) {
                    PetBlocksPlugin.logger().log(Level.WARNING, "Failed to check for updates.");
                }
            });
            NMSRegistry.registerAll();
            try {
                this.petBlockManager = new PetBlockManager(this);
                ReflectionUtils.invokeMethodByClass(PetBlocksApi.class, "initialize", new Class[]{PetMetaController.class, PetBlockController.class, GUIService.class}, new Object[]{this.petBlockManager.getPetMetaController(), this.petBlockManager.getPetBlockController(), this.guiService});
                Bukkit.getServer().getConsoleSender().sendMessage(PREFIX_CONSOLE + ChatColor.GREEN + "Enabled PetBlocks " + this.getDescription().getVersion() + " by Shynixn");
            } catch (final NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
                PetBlocksPlugin.logger().log(Level.WARNING, "Failed to enable plugin.", e);
            }
        }
    }

    /**
     * Loads the default config and saves it to the plugin folder.
     */
    @Override
    public void saveDefaultConfig() {
        try (final InputStream inputStream = this.getResource("assets/petblocks/config.yml")) {
            if (!this.getDataFolder().exists()) {
                this.getDataFolder().mkdir();
            }

            final File configFile = new File(this.getDataFolder(), "config.yml");
            if (configFile.exists()) {
                return;
            }

            try (final OutputStream outputStream = new FileOutputStream(configFile)) {
                IOUtils.copy(inputStream, outputStream);
            }
        } catch (final IOException e) {
            logger().log(Level.WARNING, "Failed to save default config.", e);
        }
    }

    /**
     * Disables the plugin PetBlocks.
     */
    @Override
    public void onDisable() {
        if (!this.disabled) {
            try {
                NMSRegistry.unregisterCustomEntities();
                this.petBlockManager.close();
            } catch (final Exception e) {
                PetBlocksPlugin.logger().log(Level.WARNING, "Failed to disable petblocks.", e);
            }
        }
    }

    /**
     * Returns the logger of the petblocks plugin.
     *
     * @return logger
     */
    public static Logger logger() {
        return logger;
    }
}

File: petblocks-bukkit-plugin/src/main/kotlin/com/github/shynixn/petblocks/bukkit/PetBlocksPlugin.kt
package com.github.shynixn.petblocks.bukkit

import com.github.shynixn.mccoroutine.bukkit.SuspendingJavaPlugin
import com.github.shynixn.mccoroutine.bukkit.setSuspendingExecutor
import com.github.shynixn.mccoroutine.bukkit.setSuspendingTabCompleter
import com.github.shynixn.mcutils.common.ChatColor
import com.github.shynixn.mcutils.common.ConfigurationService
import com.github.shynixn.mcutils.common.Version
import com.github.shynixn.mcutils.common.reloadTranslation
import com.github.shynixn.mcutils.database.api.CachePlayerRepository
import com.github.shynixn.mcutils.database.api.PlayerDataRepository
import com.github.shynixn.mcutils.database.api.SqlConnectionService
import com.github.shynixn.petblocks.bukkit.contract.PetService
import com.github.shynixn.petblocks.bukkit.contract.PetTemplateRepository
import com.github.shynixn.petblocks.bukkit.impl.commandexecutor.PetBlocksCommandExecutor
import com.github.shynixn.petblocks.bukkit.impl.listener.PetListener
import com.google.inject.Guice
import com.google.inject.Injector
import org.bukkit.Bukkit
import java.util.logging.Level

class PetBlocksPlugin : SuspendingJavaPlugin() {
    companion object {
        val prefix: String = ChatColor.BLUE.toString() + "[PetBlocks] " + ChatColor.WHITE
        val languageFiles = arrayListOf("en_us")
    }

    private var injector: Injector? = null

    /**
     * Called when this plugin is enabled
     */
    override suspend fun onEnableAsync() {
        Bukkit.getServer().consoleSender.sendMessage(prefix + ChatColor.GREEN + "Loading PetBlocks ...")
        this.saveDefaultConfig()

        if (!Version.serverVersion.isCompatible(
                Version.VERSION_1_18_R2
            )
        ) {
            Bukkit.getServer().consoleSender.sendMessage(ChatColor.RED.toString() + "================================================")
            Bukkit.getServer().consoleSender.sendMessage(ChatColor.RED.toString() + "PetBlocks does not support your server version")
            Bukkit.getServer().consoleSender.sendMessage(ChatColor.RED.toString() + "Install v" + Version.VERSION_1_18_R2.id + " - v" + Version.VERSION_1_18_R2.id)
            Bukkit.getServer().consoleSender.sendMessage(ChatColor.RED.toString() + "Plugin gets now disabled!")
            Bukkit.getServer().consoleSender.sendMessage(ChatColor.RED.toString() + "================================================")
            Bukkit.getPluginManager().disablePlugin(this)
            return
        }

        // Guice
        this.injector = Guice.createInjector(PetBlocksDependencyInjectionBinder(this))
        this.reloadConfig()

        // Register Listeners
        Bukkit.getPluginManager().registerEvents(resolve(PetListener::class.java), this)

        // Register CommandExecutors
        val configurationService = resolve(ConfigurationService::class.java)
        val petBlocksCommandExecutor = resolve(PetBlocksCommandExecutor::class.java)
        val mcTennisCommand = this.getCommand("petblocks")!!
        mcTennisCommand.aliases = configurationService.findValue("commands.petblocks.aliases")
        mcTennisCommand.usage = configurationService.findValue("commands.petblocks.usage")
        mcTennisCommand.description = configurationService.findValue("commands.petblocks.description")
        mcTennisCommand.permissionMessage = configurationService.findValue("commands.petblocks.permission-message")
        mcTennisCommand.setSuspendingExecutor(petBlocksCommandExecutor)
        mcTennisCommand.setSuspendingTabCompleter(petBlocksCommandExecutor)

        val language = configurationService.findValue<String>("language")
        this.reloadTranslation(language, PetBlocksLanguage::class.java, *languageFiles.toTypedArray())
        logger.log(Level.INFO, "Loaded language file $language.properties.")

        val sqlConnectionService = resolve(SqlConnectionService::class.java)
        sqlConnectionService.connect()
        val playerDataRepository = resolve(PlayerDataRepository::class.java)
        playerDataRepository.createIfNotExist()
        val templateRepository = resolve(PetTemplateRepository::class.java)
        templateRepository.copyTemplatesIfNotExist()

        Bukkit.getServer().consoleSender.sendMessage(prefix + ChatColor.GREEN + "Enabled PetBlocks " + this.description.version + " by Shynixn")
    }

    /**
     * Called when this plugin is disabled.
     */
    override suspend fun onDisableAsync() {
        val petService = resolve(PetService::class.java)
        petService.close()

        val playerDataRepository = resolve(CachePlayerRepository::class.java)
        playerDataRepository.clearCache()

        val sqlConnectionService = resolve(SqlConnectionService::class.java)
        sqlConnectionService.close()
    }

    /**
     * Gets a business logic from the plugin.
     * All types in the service package can be accessed.
     * Throws a [IllegalArgumentException] if the service could not be found.
     */
    private fun <S> resolve(service: Class<S>): S {
        try {
            return this.injector!!.getBinding(service).provider.get() as S
        } catch (e: Exception) {
            throw IllegalArgumentException("Service ${service.name} could not be resolved.", e)
        }
    }
}
